pkg/controller/master/addon/addon.go

```diff
@@ -109,47 +121,191 @@ func (h *Handler) MonitorChanges(key string, aObj *harvesterv1.Addon) (*harveste
 }
 
 // MonitorAddon will track the deployment of Addon
-func (h *Handler) MonitorAddon(key string, aObj *harvesterv1.Addon) (*harvesterv1.Addon, error) {
+func (h *Handler) MonitorAddonEnabled(key string, aObj *harvesterv1.Addon) (*harvesterv1.Addon, error) {
 	if aObj == nil || aObj.DeletionTimestamp != nil {
 		return nil, nil
 	}
 
-	a := aObj.DeepCopy()
+	if !aObj.Spec.Enabled {
+		return aObj, nil
+	}
 
-	if a.Status.Status != harvesterv1.AddonEnabled {
-		return a, nil
+	// if AddonDeployed or AddonFailed, return
+	if aObj.Status.Status != harvesterv1.AddonEnabled {
+		// logrus.Infof("addon %s status is", aObj.Name, aObj.Status.Status)
+		return aObj, nil
 	}
 
-	hc, err := h.helm.Cache().Get(a.Namespace, a.Name)
+	hc, err := h.helm.Cache().Get(aObj.Namespace, aObj.Name)
 	if err != nil {
-		return a, fmt.Errorf("error querying helmchart in checkHelmChartExists: %v", err)
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("waiting for helmchart %s to be created", aObj.Name)
+			h.addon.EnqueueAfter(aObj.Namespace, aObj.Name, 5*time.Second)
+			return aObj, nil
+		}
+
+		return aObj, fmt.Errorf("error querying helmchart %v", err)
 	}
 
 	if hc.Status.JobName == "" {
-		return a, fmt.Errorf("waiting for jobname to be populated in HelmChart status")
+		logrus.Infof("waiting for create jobname to be populated in helmchart status")
+		h.addon.EnqueueAfter(aObj.Namespace, aObj.Name, 5*time.Second)
+		return aObj, nil
 	}
 
-	job, err := h.job.Get(a.Namespace, hc.Status.JobName, metav1.GetOptions{})
+	if strings.HasPrefix(hc.Status.JobName, "helm-delete-") {
+		logrus.Infof("the fetched job %s is from last time helmdelete, wait install job", hc.Status.JobName)
+		h.addon.EnqueueAfter(aObj.Namespace, aObj.Name, 5*time.Second)
+		return aObj, nil
+	}
+
+	job, err := h.job.Get(aObj.Namespace, hc.Status.JobName, metav1.GetOptions{})
 	if err != nil {
-		return a, fmt.Errorf("error fetching job %s: %v", job.Name, err)
+		logrus.Infof("waiting for job %s to start", job.Name)
```

Comment:
we should refine this check to requeue if job doesnt exist, and error if there was an actual error interacting with the apiserver